### PR TITLE
Add check for uninitialized `_sync_dir` in DDP Plugin to avoid errors during error handling

### LINKED ISSUE #9263: Reconciliate processes masks errors in DDP plugin if `_sync_dir` isn't initialized
## 🐛 Bug

 If someone has no trainable parameters in their model, and call `trainer.test()`, Lightning would raise an exception because the DDP module construction would raise an exception around no parameters with `requires_grad` found.

this error is compounded by the process reconciliation check. when `configure_ddp` raises an exception, we don't call `_share_information_to_prevent_deadlock` in the DDP plugin
```
def pre_dispatch(self):
        # move the model to the correct device
        self.model_to_device()

        if self.sync_batchnorm:
            self.model = self.configure_sync_batchnorm(self.model)

       # raise exception here because on trainable params
        self.configure_ddp()

        # sync_dir initialize here, will be skipped
        self._share_information_to_prevent_deadlock()`
```

as a result, `sync_dir` is not initialized.

As part of the exception handling, we call `reconciliate_processes`: https://github.com/PyTorchLightning/pytorch-lightning/blob/35876bb75f27eb8f44220afd4bfa757a0432d233/pytorch_lightning/trainer/trainer.py#L519

which causes this path creation to fail:
```
def reconciliate_processes(self, trace: str):
        if self.world_size < 2:
            return

        sync_dir = self._sync_dir

        # The cluster may be configured to periodically purge the `/tmp`
        # directory, in which case `sync_dir` may not exist anymore at this
        # point. Idempotently create it to ensure its existence.
        Path(sync_dir).mkdir(parents=True, exist_ok=True)`
```

@ananthsub pointed out this in https://github.com/PyTorchLightning/pytorch-lightning/pull/9096

### Expected behavior

1. Initialize `sync_dir` before `configure_ddp` is called
2. Check `sync_dir` is not `None` before create path
### Environment

- PyTorch Lightning Version (e.g., 1.3.0):
- PyTorch Version (e.g., 1.8)
- Python version:
- OS (e.g., Linux):
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### PR DESCRIPTION
**Pull request by four4fish**
## What does this PR do?

python -m pytest -v tests/trainer.  passed
python -m pytest -v tests/plugins.  passed
python -m pytest -v tests/accelerators.   passed

Fixes #9263 

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by four4fish** [comment]
@ananthsub

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 1]
```diff
@@ -380,11 +380,11 @@ def pre_dispatch(self):
         if self.sync_batchnorm:
             self.model = self.configure_sync_batchnorm(self.model)
 
-        self.configure_ddp()
-
         # share ddp pids to all processes
         self._share_information_to_prevent_deadlock()
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/ddp.py:384]: could this happen at the very beginning of `pre_dispatch` ? just in case model_to_device or configure sync batchnorm raise errors

File: CHANGELOG.md [hunk 2]
```diff
@@ -267,6 +267,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed bug where data-loading functions where not getting the correct running stage passed ([#8858](https://github.com/PyTorchLightning/pytorch-lightning/pull/8858))
 
 
+- Fixed reconciliate processes masks errors in DDP plugin if _sync_dir isn't initialized ([#9267](https://github.com/PyTorchLightning/pytorch-lightning/pull/9267))
+
```
Comment (ananthsub) [inline CHANGELOG.md:271]:
```suggestion
- Fixed error handling in DDP process reconciliation when `_sync_dir` was not initialized ([#9267](https://github.com/PyTorchLightning/pytorch-lightning/pull/9267))

```

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 3]
```diff
@@ -480,6 +480,10 @@ def reconciliate_processes(self, trace: str):
 
         sync_dir = self._sync_dir
 
+        if not sync_dir:
+            rank_zero_warn("pid is not shared between processes in DDP and sync_dir is not initialized")
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/ddp.py:484]: Can we improve the warning message for users who have no idea about DDP or deadlocks or PIDs?

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 3 above)
Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/ddp.py:484]:
```suggestion
            rank_zero_warn("Error handling mechanism for deadlock detection is uninitialized. Skipping check.")
```
?

**Review by awaelchli — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 4]
```diff
@@ -374,6 +374,9 @@ def determine_ddp_device_ids(self):
         return [self.root_device.index]
 
     def pre_dispatch(self):
+        # share ddp pids to all processes
+        self._share_information_to_prevent_deadlock()
```
[the remark this answers is not in the document]
Comment (four4fish) [reply pytorch_lightning/plugins/training_type/ddp.py:378]: Yeah, for for the plugins extend ddp and overrides pre_dispatch, the _share_information_to_prevent_deadlock is not called and sync_dir is None. Does that means the reconciliate_processes never worked for these plugins?

[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/ddp.py:378]: > Does that means the reconciliate_processes never worked for these plugins?

it appears not 😞

[the remark this answers is not in the document]
Comment (four4fish) [reply pytorch_lightning/plugins/training_type/ddp.py:378]: created issue for this https://github.com/PyTorchLightning/pytorch-lightning/issues/9291
